pkg/controller/aerospikecluster/aerospikecluster_controller.go

```diff
@@ -131,91 +137,246 @@ func (r *ReconcileAerospikeCluster) Reconcile(request reconcile.Request) (reconc
+	// TODO: Replace the sleep.
+	time.Sleep(time.Second * 2)
+
+	return nil
+}
+
+func (r *ReconcileAerospikeCluster) isNewCluster(aeroCluster *aerospikev1alpha1.AerospikeCluster) (bool, error) {
+	if aeroCluster.Status.AerospikeConfig != nil {
+		// We have valid status, cluster cannot be new.
+		return false, nil
 	}
 
-	if aeroCluster.Status.AerospikeConfig == nil {
-		// Clearly a failed cluster create since we never set Status to nil after successful cluster create.
-		return r.recoverFailedCreate(aeroCluster, found)
+	statefulSetList, err := r.getClusterStatefulSets(aeroCluster)
+
+	if err != nil {
+		return false, err
 	}
 
-	return r.reconcileCluster(aeroCluster, found)
+	// Cluster can have status nil and still have pods on failures.
+	// For cluster to be new there should be no pods in the cluster.
+	return len(statefulSetList.Items) == 0, nil
 }
 
-func (r *ReconcileAerospikeCluster) createCluster(aeroCluster *aerospikev1alpha1.AerospikeCluster) (reconcile.Result, error) {
+func (r *ReconcileAerospikeCluster) hasClusterFailed(aeroCluster *aerospikev1alpha1.AerospikeCluster) (bool, error) {
+	isNew, err := r.isNewCluster(aeroCluster)
+
+	if err != nil {
+		// Checking cluster status failed.
+		return false, err
+	}
+
+	return !isNew && aeroCluster.Status.AerospikeConfig == nil, nil
+}
+
+func (r *ReconcileAerospikeCluster) createSTSForCluster(aeroCluster *aerospikev1alpha1.AerospikeCluster, rackState RackState) error {
 	logger := pkglog.New(log.Ctx{"AerospikeCluster": utils.ClusterNamespacedName(aeroCluster)})
-	logger.Info("Create new Aerospike cluster")
+	logger.Info("Create new Aerospike cluster if needed")
 
+	// NoOp if already exist
+	logger.Info("AerospikeCluster", log.Ctx{"Spec": aeroCluster.Spec})
 	if err := r.createHeadlessSvc(aeroCluster); err != nil {
 		logger.Error("Failed to create headless service", log.Ctx{"err": err})
-		return reconcile.Result{}, err
+		return err
 	}
-
 	// Bad config should not come here. It should be validated in validation hook
-	if err := r.buildConfigMap(aeroCluster); err != nil {
+	cmName := getNamespacedNameForConfigMap(aeroCluster, rackState.Rack.ID)
+	if err := r.buildConfigMap(aeroCluster, cmName, rackState.Rack); err != nil {
 		logger.Error("Failed to create configMap from AerospikeConfig", log.Ctx{"err": err})
-		return reconcile.Result{}, err
+		return err
 	}
 
-	// Create an empty status object.
-	if err := r.createStatus(aeroCluster); err != nil {
-		return reconcile.Result{}, err
-	}
+	// _, err := r.createClusterSTS(aeroCluster, rackStateList)
+	// if err != nil {
+	// 	return err
+	// }
 
-	if found, err := r.createStatefulSetForAerospikeCluster(aeroCluster); err != nil {
-		logger.Error("Failed to create statefulset", log.Ctx{"err": err})
-		// return reconcile.Result{}, nil
+	found := &appsv1.StatefulSet{}
+	stsName := getNamespacedNameForStatefulSet(aeroCluster, rackState.Rack.ID)
+	found, err := r.createStatefulSetForAerospikeCluster(aeroCluster, stsName, rackState)
```

Comment:
Rename to createStatefulSetForRack

Reply:
now naming is like reconcileRacks,  reconcileRack, createRack, deleteRack, scaleUpRack, updateRack....